# Class Data and Methods - Part 2

### 0. Define a method to unset classes and objects (to help with running code blocks multiple times)

In [44]:
def unset(varname):
    """
    Unsets a global variable from memory. Helpful when we are running code blocks in Jupyter multiple times.
    """
    try:
        del globals()[varname]
        print('deleted ' + varname)
    except:
        print(varname + ' does not exist')

### 1. Use the class's name instead of self for the conn_count.

In [103]:
unset('ConnectionC')
unset('connect')
unset('connect2')
unset('connect3')
unset('connect4')
unset('connect5')

ConnectionC does not exist
connect does not exist
connect2 does not exist
connect3 does not exist
connect4 does not exist
deleted connect5


In [104]:
class ConnectionC:
    port = 55000
    conn_limit = 10
    conn_count = 0
    
    def __init__(self, host):
        # set the host for the instance
        self.host = host
        # set the port based on the class variable port
        
        # add 1 to the class's connection_count
        if ConnectionC.conn_count < ConnectionC.conn_limit:
            ConnectionC.conn_count += 1
        
        self.port = self.port + ConnectionC.conn_count

    def close(self):
        # reduce the class's connection_count by 1
        # self.conn_count -= 1
        ConnectionC.conn_count = ConnectionC.conn_count - 1
        
    def __repr__(self):
        return f"{self.host}, {self.port}"

In [105]:
connect = ConnectionC('localhost')

In [106]:
connect.__dict__

{'host': 'localhost', 'port': 55001}

In [109]:
ConnectionC.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': 0,
              '__init__': <function __main__.ConnectionC.__init__(self, host)>,
              'close': <function __main__.ConnectionC.close(self)>,
              '__repr__': <function __main__.ConnectionC.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionC' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionC' objects>,
              '__doc__': None})

The class variable port return 55000, whereas the instance variable port returns 55001

In [107]:
connect.close()

Each time I run connect.close(), it decrements conn_count. It is currently set to 0 after connect.close() was run once.
Interestingly, the class variable port still says 55000.

In [110]:
connect.close()

In [111]:
ConnectionC.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': -1,
              '__init__': <function __main__.ConnectionC.__init__(self, host)>,
              'close': <function __main__.ConnectionC.close(self)>,
              '__repr__': <function __main__.ConnectionC.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionC' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionC' objects>,
              '__doc__': None})

^ When I ran connect.close() a second time, it decremented the class variable conn_count by one, setting it to -1
What would happen if I created a second instance of the ConnectionC class?

In [113]:
connect.__dict__

{'host': 'localhost', 'port': 55001}

In [114]:
connect2 = ConnectionC('localhost')
connect2.__dict__

{'host': 'localhost', 'port': 55000}

The port went down here due to conn_count being -1.

In [115]:
ConnectionC.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': 0,
              '__init__': <function __main__.ConnectionC.__init__(self, host)>,
              'close': <function __main__.ConnectionC.close(self)>,
              '__repr__': <function __main__.ConnectionC.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionC' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionC' objects>,
              '__doc__': None})

In [116]:
connect3 = ConnectionC('localhost')
connect3.__dict__

{'host': 'localhost', 'port': 55001}

In [117]:
ConnectionC.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': 1,
              '__init__': <function __main__.ConnectionC.__init__(self, host)>,
              'close': <function __main__.ConnectionC.close(self)>,
              '__repr__': <function __main__.ConnectionC.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionC' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionC' objects>,
              '__doc__': None})

In [118]:
connect4 = ConnectionC('localhost')
connect4.__dict__

{'host': 'localhost', 'port': 55002}

In [119]:
connect2.close()
connect3.close()
connect4.close()
connect4.close()

In [120]:
ConnectionC.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': -2,
              '__init__': <function __main__.ConnectionC.__init__(self, host)>,
              'close': <function __main__.ConnectionC.close(self)>,
              '__repr__': <function __main__.ConnectionC.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionC' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionC' objects>,
              '__doc__': None})

In [121]:
connect4.__dict__

{'host': 'localhost', 'port': 55002}

In [122]:
connect5 = ConnectionC('localhost')

In [123]:
connect5.__dict__

{'host': 'localhost', 'port': 54999}

^ It's odd that the port number above is 54999. This is because I called the close() method on previous objects multiple times. Each time I ran the close() instance method, it decremented the class_count class variable. This in turn affected the calculated port when a new instance was instantiated, because in the constructor, we add the port instance variable to the class_count class variable. By adding a negative value, it subtracts from the minimum port value of 55000, instead of adding to it.
Conclusion: class variables seem to be shared across instances. While this could be useful, it could also be very complicated, confusing and buggy if not used correctly.

### 2. Working with subclasses

In [124]:
class ConnectionCa(ConnectionC):
    pass

I created a subclass of ConnectionC here

In [81]:
unset('ConnectionCa')

ConnectionCa does not exist


In [125]:
connectCa1 = ConnectionCa('localhost')
connectCa1.__dict__

{'host': 'localhost', 'port': 55000}

In [126]:
ConnectionCa.__dict__

mappingproxy({'__module__': '__main__', '__doc__': None})

In [127]:
connectCa2 = ConnectionCa('localhost')

In [128]:
connectCa2.__dict__

{'host': 'localhost', 'port': 55001}

In [129]:
ConnectionCa.__dict__

mappingproxy({'__module__': '__main__', '__doc__': None})

^ ConnectionCa (the subclass), doesn't copy the class variables from the parent (it seems to inherit them)

In [130]:
connect5.close()

^ Call the close() method on the instance defined in section 1.

In [131]:
connectCa3 = ConnectionCa('localhost')

In [132]:
connectCa3.__dict__

{'host': 'localhost', 'port': 55001}

Looks like it is inheriting the class variables from the parent class

### 3. Using "__ class __" instead of the class's hardcoded name

In [133]:
class ConnectionD:
    port = 55000
    conn_limit = 10
    conn_count = 0
    
    def __init__(self, host):
        # set the host for the instance
        self.host = host
        # set the port based on the class variable port
        
        # add 1 to the class's connection_count
        if __class__.conn_count < __class__.conn_limit:
            __class__.conn_count += 1
        
        self.port = self.port + __class__.conn_count

    def close(self):
        # reduce the class's connection_count by 1
        # self.conn_count -= 1
        __class__.conn_count = __class__.conn_count - 1
        
    def __repr__(self):
        return f"{self.host}, {self.port}"

In [134]:
connectD = ConnectionD('localhost')
connectD.__dict__

{'host': 'localhost', 'port': 55001}

It seemed to work when we used dunder __class__

In [135]:
connectD2 = ConnectionD('localhost')
connectD2.__dict__

{'host': 'localhost', 'port': 55002}

In [106]:
print(connectD)
print(connectD2)

localhost, 55001
localhost, 55002


In [136]:
ConnectionD.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': 2,
              '__init__': <function __main__.ConnectionD.__init__(self, host)>,
              'close': <function __main__.ConnectionD.close(self)>,
              '__repr__': <function __main__.ConnectionD.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionD' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionD' objects>,
              '__doc__': None})

In [137]:
connectD2.close()

In [138]:
ConnectionD.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': 1,
              '__init__': <function __main__.ConnectionD.__init__(self, host)>,
              'close': <function __main__.ConnectionD.close(self)>,
              '__repr__': <function __main__.ConnectionD.__repr__(self)>,
              '__dict__': <attribute '__dict__' of 'ConnectionD' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionD' objects>,
              '__doc__': None})

In [139]:
connectD2.__dict__

{'host': 'localhost', 'port': 55002}

In [141]:
connectD3 = ConnectionD('localhost')

In [142]:
connectD3.__dict__

{'host': 'localhost', 'port': 55002}

This is a bug I hadn't noticed before ... now connectD2 and connectD3 share the same port. But this bug is due to the usage of class variables for the connection count.

### 4. Using class methods (@classmethod decorator)

In [161]:
class ConnectionCM1:
    port = 55000
    conn_limit = 10
    conn_count = 0
    host = ''
    
    @classmethod
    def get_next_port(cls):
        return cls.port + 1
        
    @classmethod
    def add_connection(cls, host):
        cls.host = host
        
        # add 1 to the class's connection_count
        if cls.conn_count < cls.conn_limit:
            cls.conn_count += 1
        
        cls.port = cls.get_next_port()
        
    @classmethod
    def remove_connection(cls):
        # reduce the class's connection_count by 1
        cls.conn_count = cls.conn_count - 1
        cls.port = cls.port - 1
        
    @classmethod
    def get_connection_count(cls):
        return cls.conn_count
        

In [160]:
unset('ConnectionCM1')

deleted ConnectionCM1


In [162]:
ConnectionCM1.add_connection('localhost')

In [163]:
ConnectionCM1.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55001,
              'conn_limit': 10,
              'conn_count': 1,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM1.get_next_port at 0x107cb8d60>)>,
              'add_connection': <classmethod(<function ConnectionCM1.add_connection at 0x107cb8e00>)>,
              'remove_connection': <classmethod(<function ConnectionCM1.remove_connection at 0x107cb9300>)>,
              'get_connection_count': <classmethod(<function ConnectionCM1.get_connection_count at 0x107cb8fe0>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM1' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM1' objects>,
              '__doc__': None})

In [164]:
ConnectionCM1.add_connection('localhost')

In [165]:
ConnectionCM1.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55002,
              'conn_limit': 10,
              'conn_count': 2,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM1.get_next_port at 0x107cb8d60>)>,
              'add_connection': <classmethod(<function ConnectionCM1.add_connection at 0x107cb8e00>)>,
              'remove_connection': <classmethod(<function ConnectionCM1.remove_connection at 0x107cb9300>)>,
              'get_connection_count': <classmethod(<function ConnectionCM1.get_connection_count at 0x107cb8fe0>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM1' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM1' objects>,
              '__doc__': None})

In [166]:
ConnectionCM1.remove_connection()

In [167]:
ConnectionCM1.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55001,
              'conn_limit': 10,
              'conn_count': 1,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM1.get_next_port at 0x107cb8d60>)>,
              'add_connection': <classmethod(<function ConnectionCM1.add_connection at 0x107cb8e00>)>,
              'remove_connection': <classmethod(<function ConnectionCM1.remove_connection at 0x107cb9300>)>,
              'get_connection_count': <classmethod(<function ConnectionCM1.get_connection_count at 0x107cb8fe0>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM1' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM1' objects>,
              '__doc__': None})

In [168]:
ConnectionCM1.remove_connection()
ConnectionCM1.__dict__

mappingproxy({'__module__': '__main__',
              'port': 55000,
              'conn_limit': 10,
              'conn_count': 0,
              'host': 'localhost',
              'get_next_port': <classmethod(<function ConnectionCM1.get_next_port at 0x107cb8d60>)>,
              'add_connection': <classmethod(<function ConnectionCM1.add_connection at 0x107cb8e00>)>,
              'remove_connection': <classmethod(<function ConnectionCM1.remove_connection at 0x107cb9300>)>,
              'get_connection_count': <classmethod(<function ConnectionCM1.get_connection_count at 0x107cb8fe0>)>,
              '__dict__': <attribute '__dict__' of 'ConnectionCM1' objects>,
              '__weakref__': <attribute '__weakref__' of 'ConnectionCM1' objects>,
              '__doc__': None})